File: Data Structures & Algorithms/last-stone-weight/submission-0.go
func lastStoneWeight(stones []int) int {

    for len(stones) > 1 {
        n := len(stones)

        quickSort(stones, 0 , n-1 )
         
        x := stones[0]
        y := stones[1]

        stones = stones[2:]

        if x != y {
            stones = append(stones, x-y)
        }
    }

    if len(stones) == 0 {
        return 0
    }

    return stones[0]
}


func quickSort(nums []int , low , high int) {
    if low < high {
        p := partitionAsc(nums , low ,high)
        quickSort(nums , low , p-1)
        quickSort(nums , p+1 , high)
    }
}

func partitionAsc(nums []int , low , high int) int {
    pivot := nums[high]
    i := low

    for j := low ; j < high ; j++ {
        if nums[j] >= pivot {
            nums[i] , nums[j] = nums[j] , nums[i]
            i++
        }
    }

    nums[i] , nums[high] = nums[high] , nums[i]

    return i
}

File: Data Structures & Algorithms/sort-colors/submission-0.py
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        self.quickSort(nums , 0, len(nums) - 1)
    
    def quickSort(self , nums , low,  high):
        if low < high:
            pivot_index = self.partition(nums,low,high)
            self.quickSort(nums,low, pivot_index -1)
            self.quickSort(nums, pivot_index+1 , high  )

    
    def partition(self , nums, low, high):
        pivot = nums[low]
        i = low+1

        for j in range(low+1 , high+1):
            if nums[j] <= pivot:
                nums[i], nums[j] = nums[j] , nums[i]
                i += 1
        
        nums[low], nums[i - 1] = nums[i - 1], nums[low]
        return i - 1


        